modules/host/contractcreation.go

```diff
@@ -2,258 +2,250 @@ package host
-			return
-		}
-		contract := txn.FileContracts[0]
-
-		// Check that the contained FileContract fits host criteria for taking
-		// files, replying with the error if there's a problem.
-		h.mu.Lock()
-		txn, err = h.considerContract(txn, startBlock)
-		h.mu.Unlock()
-		if err != nil {
-			_, err = encoding.WriteObject(conn, err.Error())
-			return
-		}
-		_, err = encoding.WriteObject(conn, modules.AcceptContractResponse)
-		if err != nil {
-			return
-		}
-
-		// Create file.
-		h.mu.Lock()
-		filename := h.nextFilename()
-		fullname := filepath.Join(h.hostDir, filename)
-		h.mu.Unlock()
-		file, err := os.Create(fullname)
-		if err != nil {
-			return
-		}
-		defer file.Close()
-
-		// If there's an error upon return, delete the file that's been created.
-		defer func() {
-			if err != nil {
-				_, err = encoding.WriteObject(conn, err.Error())
-				os.Remove(fullname)
-				h.mu.Lock()
-				h.spaceRemaining -= int64(contract.FileSize)
-				h.mu.Unlock()
-			}
-		}()
-
-		// Download file contents.
-		_, err = io.CopyN(file, conn, int64(contract.FileSize))
-		if err != nil {
-			return
-		}
-
-		// Check that the file matches the merkle root in the contract.
-		_, err = file.Seek(0, 0)
-		if err != nil {
-			return
-		}
-		merkleRoot, err := hash.ReaderMerkleRoot(file, hash.CalculateSegments(contract.FileSize))
+	// Read the contract terms.
+	var terms modules.ContractTerms
+	err = encoding.ReadObject(conn, &terms, maxContractLen)
+	if err != nil {
+		return
+	}
+
+	// Consider the contract terms. If they are unnacceptable, return an error
+	// describing why.
+	h.mu.RLock()
+	err = h.considerTerms(terms)
+	h.mu.RUnlock()
+	if err != nil {
+		_, err = encoding.WriteObject(conn, err.Error())
+		return
+	}
+
+	// terms are acceptable; allocate space for file
+	h.mu.Lock()
+	file, path, err := h.allocate(terms.FileSize)
+	h.mu.Unlock()
+	if err != nil {
+		return
+	}
+	defer file.Close()
+
+	// rollback everything if something goes wrong
+	defer func() {
 		if err != nil {
-			return
-		}
-		if merkleRoot != contract.FileMerkleRoot {
-			err = errors.New("uploaded file has wrong merkle root")
-			return
-		}
-
-		// Check that the file arrived in time.
-		if h.state.Height() >= contract.Start-2 {
-			err = errors.New("file not uploaded in time, refusing to go forward with contract")
-			return
-		}
-
-		// Put the contract in a list where the host will be performing proofs of
-		// storage.
-		h.mu.Lock()
-		h.contracts[txn.FileContractID(0)] = contractObligation{
-			filename: filename,
-		}
-		h.mu.Unlock()
-		fmt.Println("Accepted contract")
-
-		// Submit the transaction.
-		h.state.AcceptTransaction(txn)
-	*/
+			h.deallocate(terms.FileSize, path)
+		}
+	}()
+
+	// signal that we are ready to download file
+	_, err = encoding.WriteObject(conn, modules.AcceptContractResponse)
+	if err != nil {
+		return
+	}
+
+	// file transfer is going to take a while, so extend the timeout.
+	// This assumes a minimum transfer rate of ~1 Mbps
```

Comment:
This is a temporary fix anyway. I imagine 1Mbps will be quite generous during in-house testing.